File: 백준/Gold/9019. DSLR/DSLR.py
from collections import deque
import sys
input = sys.stdin.readline

cmd_map = {0: 'D', 1: 'S', 2: 'L', 3: 'R'}

def cal(x, i):
    if i == 0:
        return (x*2) % 10000
    elif i == 1:
        return (x-1) % 10000
    elif i == 2:
        # 1234 -> 2341
        return (x % 1000) * 10 + x//1000
    else:
        # 1234 -> 4123
        return (x % 10) * 1000 + x//10

for _ in range(int(input())):
    a, b = map(int, input().split())
    
    q = deque()
    q.append((a, ""))
    visited = [False] * 10_001
    visited[a] = True
    
    while q:
        x, cmd = q.popleft()
        
        if x == b:
            print(cmd)
            break
        
        for i in range(4):
            nx = cal(x, i)
            
            if not visited[nx]:
                visited[nx] = True
                q.append((nx, cmd+cmd_map[i]))
    

File: 백준/Gold/9019. DSLR/DSLR.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.*;

class Main {

    static Map<Integer, String> cmdMap = Map.of(0, "D", 1, "S", 2, "L", 3, "R");
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());

        while (T -- > 0) {
            String[] arr = br.readLine().split(" ");
            int a = Integer.parseInt(arr[0]);
            int b = Integer.parseInt(arr[1]);

            Queue<Tuple> q = new LinkedList<>();
            q.add(new Tuple(a, ""));
            boolean[] visited = new boolean[10000];
            visited[a] = true;

            while (!q.isEmpty()) {
                Tuple now = q.poll();

                if (now.num == b) {
                    System.out.println(now.cmd);
                    break;
                }

                for (int i = 0; i < 4; i++) {
                    int newNum = process(i, now.num);
                    if (! visited[newNum]) {
                        visited[newNum] = true;
                        q.add(new Tuple(newNum, now.cmd+cmdMap.get(i)));
                    }
                }
            }

        }

    }

    private static int process(int i, int x) {
        if (i == 0)
            return (x * 2) % 10000;
        if (i == 1)
            return x == 0 ? 9999 : x - 1;
        if (i == 2)     // 1234 -> 2341 
            return (x % 1000) * 10 + x / 1000;
        // 1234 -> 4123
        return (x % 10) * 1000 + x / 10;
    }
}

class Tuple {
    int num;
    String cmd;

    public Tuple(int num, String cmd) {
        this.num = num;
        this.cmd = cmd;
    }
}